README.md
# Vertica output plugin for Embulk

## Overview

* **Plugin type**: output
* **Resume supported**: no
* **Cleanup supported**: yes
* **Dynamic table creating**: yes

## Configuration

- **host**: hostname (string, default: localhost)
- **port**: port number (integer, default: 5433)
- **user**: user name (string, required)
- **password**: password (string, default: '')
- **database**: database name (string, default: vdb)
- **schema**:   schema name (string, default: public)
- **table**:    table name (string, required)
- **mode**:     "insert", or "replace". See bellow. (string, default: insert)
- **copy_mode**: specifies how data is loaded into the database. See vertica documents for details. (`AUTO`, `DIRECT`, or `TRICKLE`. default: `AUTO`)
- **pool**: number of output threads, this number controls number of concurrency to issue COPY statements (integer, default: processor_count, that is, number of threads in input plugin)
- **abort_on_error**: stops the COPY command if a row is rejected and rolls back the command. No data is loaded. (bool, default: false)
- **compress**: compress input (`GZIP`, or `UNCOMPRESSED`, default: `UNCOMPRESSED`)
- **reject_on_materialized_type_error**: uses `reject_on_materialized_type_error` option for fjsonparser(). This rejects rows if any of column types and value types do not fit, ex) double value into INT column fails. See vertica documents for details. (bool, default: false)
- **default_timezone**: the default timezone for column_options (string, default is "UTC")
- **column_options**: advanced: a key-value pairs where key is a column name and value is options for the column.
  - **type**: type of a column when this plugin creates new tables such as `VARCHAR(255)`, `INTEGER NOT NULL UNIQUE`. This is used on creating intermediate tables (insert and truncate_insert modes) and on creating a new target table. (string, default: depends on input column type, see below)
    - boolean:   `BOOLEAN`
    - long:      `INT` (same with `BIGINT` in vertica)
    - double:    `FLOAT` (same with `DOUBLE PRECISION` in vertica)
    - string:    `VARCHAR`
    - timestamp: `TIMESTAMP`
  - **value_type**:  The types (embulk types) of values to convert (string, default: no conversion. See below for available types)
    - boolean:   `boolean`, `string`
    - long:      `boolean`, `long`, `double`, `string`, `timestamp`
    - double:    `boolean`, `long`, `double`, `string`, `timestamp`
    - string:    `boolean`, `long`, `double`, `string`, `timestamp`
    - timestamp: `boolean`, `long`, `double`, `string`, `timestamp`
  - **timestamp_format**: timestamp format to convert into/from `timestamp` (string, default is "%Y-%m-%d %H:%M:%S %z")
  - **timezone**: timezone to convert into/from `timestamp` (string, default is `default_timezone`).
- **json_payload**: Assuming first columns of records are json string, COPY each json into vertica directly using fjsonparser. This would improve performance by avoiding construction of json in this jruby plugin. ToDo: auto table creation is not supported for this mode yet (bool, default is false)
- **resource_pool**: Configure resource pool (string, default is nil)

### Modes

* **insert**:
  * Behavior: This mode copies rows to an intermediate table first. If all those tasks run correctly, runs `INSERT INTO <target_table> SELECT * FROM <intermediate_table>`
* **replace**:
  * Behavior: This mode copies rows to an intermediate table first. If all those tasks run correctly, swaps the target table and the intermediate table, and drops the old target table
* **drop_insert**:
  * Behavior: Same with insert mode excepting that it drops the target table first. This is aimed to use for development.

## Example

```yaml
out:
  type: vertica 
  host: 127.0.0.1
  user: dbadmin
  password: xxxxxxx
  database: vdb
  schema: sandbox
  table: embulk_test
  copy_mode: DIRECT
  abort_on_error: true
  column_options:
    id:   {type: INT}
    name: {type: VARCHAR(255)}
    date: {type: DATE, value_type: timestamp, timezone: "+09:00"}
```

## Development

Run example:

```
$ embulk bundle install --path vendor/bundle
$ embulk -J-O -R--dev run -b . -l debug example/example.yml
```

Release gem:

```
$ bundle exec rake release
```

## ChangeLog

[CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# 0.7.4 (2016/05/07)

Fixes:

* Add missing "require 'timeout'"

# 0.7.3 (2016/03/27)

Enhancements:

* Add write_timeout option
* Add finish_timeout option
* Add dequeuen_timeout option

Changes:

* Transfer repository from eratostennis to sonots

# 0.7.2 (2016/02/10)

Fixes:

* Fix to get stuck if no input is comming
* Fix to get stcuk if one record is longer thant IO pipe maximum size (64k bytes for Linux, and Mac OSX)

# 0.7.1 (2016/01/30)

Enhancements:

* Create table with INCLUDING PROJECTIONS on mode `REPLACE`

# 0.7.0 (2016/01/29)

Breaking Changes:

* rename mode `REPLACE` to `DROP_INSERT`

Enhancements:

* Add mode `REPLACE` which copies rows to an intermediate table first. If all those tasks run correctly, swaps the target table and the intermediate table, and drops the old target table

# 0.6.2 (2016/01/28)

Enhancements:

* Use `CREATE TABLE LIKE` statement to create a temp table to copy nullable or other parameters

# 0.6.1 (2016/01/26)

Fixes:

* Remove debug code...

# 0.6.0 (2016/01/26)

Changes:

* Add `/*+ direct */` hint to insert select if copy_mode == DIRECT

# 0.5.9 (2016/01/26)

Enhancements:

* log speed (rows/sec) of progress

Trivial changes:

* Use ::Jvertica.quote to quote resource_pool
* change log_level of push / pop finish
* change log_level of select result from debug to trace

# 0.5.8 (2016/01/24)

Enhancements:

* Add resource_pool option

# 0.5.7 (2016/01/22)

Enhancements:

* Add json_payload option to avoid construction of JSON in this jruby plugin for performance improvement

# 0.5.6 (2016/01/22)

Enhancements:

* Perform to_json in embulk threads. This achieves better performance when pool < num_threads

# 0.5.5 (2016/01/22)

Changes:

* Show input_num_rows info log to see progress

# 0.5.4 (2016/01/21)

Enhancements:

* Log rejected record

# 0.5.3 (2016/01/09)

Enhancements:

* Add more trace logs

# 0.5.2 (2016/01/09)

Fixes:

* Add nil check for timestamp converter
* Fix error handling of output thread

# 0.5.1 (2015/12/04)

Fixes:

* Need mutex.synchroize for OutputThreadPool#enqueue because it is called by #add which is ran by multiple threads

# 0.5.0 (2015/12/04)

Changes:

* Use thread pool instead of connection pool #13

# 0.4.1 (2015/12/04)

Fixes:

* Create internal vertica projection beforehand to avoid S Lock error 

# 0.4.0 (2015/11/24)

Enhancements:

* Support connection pool

# 0.3.1 (2015/11/20)

Fixes:

* Fix timezone support for the case that column_options is not specified (use default_timezone)

# 0.3.0 (2015/11/17)

Changes:

* Change log level of COMMIT statement from info to debug

# 0.2.9 (2015/11/17)

Changes:

* Change log level of COPY statement from info to debug

# 0.2.8 (2015/11/06)

Enhancements:

* Get sql schema from the existing target table to create internal temporary tables to avoid schema conflicts

# 0.2.7 (2015/11/06)

Skipped

# 0.2.6 (2015/11/06)

Fixes:

* Fix not to raise ConfigError for upcase mode, and copy_mode

# 0.2.5 (2015/10/26)

Changes:

* Output task_reports log as json

# 0.2.4 (2015/10/23)

Changes:

* Rename `username` to `user` to be compatible with ruby vertica gem and jruby jvertica gem
  * still, support `username` for backward compatibility

# 0.2.3 (2015/09/16)

Changes:

* Commit all pages at burst (in each task)

Enhancements:

* Return task_reports

# 0.2.2 (2015/07/24)

Changes:

* Change some log level from debug to info

# 0.2.1 (2015/07/24)

Fixes:

* Fix to support timezone 'UTC'

# 0.2.0 (2015/07/24)

Enhancements:

* Add `default_timezone` option

# 0.1.9 (2015/07/24)

Enhancements:

* Support `timezone` for string converter

# 0.1.8 (2015/07/24)

Enhancements:

* Support `value_type`, `timezone_format`, `timezone` option for column_options

# 0.1.7 (2015/07/24)

Enhancements:

* Add `reject_on_materialized_type_error` option

# 0.1.6 (2015/07/23)

Enhancements:

* Enhancement of debug log

# 0.1.5 (2015/07/23)

Fixes:

* Use PARSER fjsonparser() instead of DELIMITER ',', otherwise escape is too difficult to do

# 0.1.4 (2015/07/10)

Fixes:

* Just fix gemspec

# 0.1.3 (2015/07/05)

Enhancements:

* Escape schama, table names
* Add `abort_on_error` option

# 0.1.2 (2015/06/26)

Enhancements:

* Add `column_options` option

# 0.1.1 (2015/06/25)

Enhancements:

* Add `copy_mode` option
* Use `jvertica` gem instead of `vertica` gem

# 0.1.0

first version



lib/embulk/output_vertica.rb
module Embulk
  module Plugin
    module Vertica

      class NotSupportedSchema < Exception; end
      class OutputVertica < OutputPlugin

        require 'vertica'
        Plugin.register_output('vertica', self)

        def self.transaction(config, schema, processor_count, &control)
          task = {
            'host' => config.param('host', :string, :default => 'localhost'),
            'port' => config.param('port', :integer, :default => 5433),
            'username' => config.param('username', :string),
            'password' => config.param('password', :string, :default => ''),
            'database' => config.param('database', :string, :default => 'vdb'),
            'schema' => config.param('schema', :string, :default => 'public'),
            'table' => config.param('table', :string),
          }

          now = Time.now
          unique_name = "%08x%08x" % [now.tv_sec, now.tv_nsec]
          task['temp_table'] = "#{task['table']}_LOAD_TEMP_#{unique_name}"

          sql_schema = self.to_vertica_schema schema

          connect(task) do |jv|
            # drop table if exists "DEST"
            # 'create table if exists "TEMP" ("COL" json)'
            jv.query %[drop table if exists #{task['schema']}.#{task['temp_table']}]
            jv.query %[create table #{task['schema']}.#{task['temp_table']} (#{sql_schema})]
          end

          begin
            yield(task)
            connect(task) do |jv|
              # create table if not exists "DEST" ("COL" json)
              # 'insert into "DEST" ("COL") select "COL" from "TEMP"'
              jv.query %[create table if not exists #{task['schema']}.#{task['table']} (#{sql_schema})]
              jv.query %[insert into #{task['schema']}.#{task['table']} select * from #{task['schema']}.#{task['temp_table']}]
              jv.query %[COMMIT]
            end
          ensure
            connect(task) do |jv|
              # 'drop table if exists TEMP'
              jv.query %[drop table if exists #{task['schema']}.#{task['temp_table']}]
            end
          end
          return {}
        end

        def self.connect(task)
          jv = ::Vertica.connect({
            host: task['host'],
            port: task['port'],
            user: task['username'],
            password: task['password'],
            database: task['database'],
          })

          if block_given?
            begin
              yield jv
            ensure
              jv.close
            end
          end
          jv
        end

        def self.to_vertica_schema(schema)
            schema.names.zip(schema.types)
            .map { |name, type| "#{name} #{to_sql_type(type)}" }
            .join(',')
        end

         def self.to_sql_type(type)
           case type
             when :boolean then 'BOOLEAN'
             when :long then 'INT'
             when :double then 'FLOAT'
             when :string then 'VARCHAR'
             when :timestamp then 'TIMESTAMP'
             else fail NotSupportedSchema, "embulk-plugin-output-vertica cannot take column type #{type}"
           end
         end

        def initialize(task, schema, index)
          super
          @jv = self.class.connect(task)
        end

        def close
          @jv.close
        end

        def add(page)
          sql = "COPY #{@task['schema']}.#{@task['temp_table']} FROM STDIN DELIMITER ','"
          @jv.copy(sql) do |stdin|
            page.each_with_index do |record, idx|
              stdin << record.map {|v| ::Vertica.quote(v)}.join(",")
              stdin << "\n" unless record.size-1 == idx
            end
          end
        end

        def finish
        end

        def abort
        end

        def commit
          {}
        end
      end
    end
  end
end


lib/embulk/output/vertica/output_thread.rb
require 'zlib'
require 'timeout'

module Embulk
  module Output
    class Vertica < OutputPlugin
      class CommitError < ::StandardError; end
      class TimeoutError < ::Timeout::Error; end
      class DequeueTimeoutError < TimeoutError; end
      class FinishTimeoutError < TimeoutError; end
      class WriteTimeoutError < TimeoutError; end

      $embulk_output_vertica_thread_dumped = false

      class OutputThreadPool
        def initialize(task, schema, size)
          @task = task
          @size = size
          @schema = schema
          @converters = ValueConverterFactory.create_converters(schema, task['default_timezone'], task['column_options'])
          @output_threads = size.times.map { OutputThread.new(task) }
          @current_index = 0
        end

        def enqueue(page)
          json_page = []
          page.each do |record|
            json_page << to_json(record)
          end
          @mutex.synchronize do
            @output_threads[@current_index].enqueue(json_page)
            @current_index = (@current_index + 1) % @size
          end
        end

        def start
          @mutex = Mutex.new
          @size.times.map {|i| @output_threads[i].start }
        end

        def commit
          Embulk.logger.debug "embulk-output-vertica: commit"
          task_reports = @mutex.synchronize do
            @size.times.map {|i| @output_threads[i].commit }
          end
          unless task_reports.all? {|task_report| task_report['success'] }
            raise CommitError, "some of output_threads failed to commit"
          end
          task_reports
        end

        def to_json(record)
          if @task['json_payload']
            record.first
          else
            Hash[*(@schema.names.zip(record).map do |column_name, value|
              [column_name, @converters[column_name].call(value)]
            end.flatten!(1))].to_json
          end
        end
      end

      class OutputThread
        def initialize(task)
          @task = task
          @queue = SizedQueue.new(1)
          @num_input_rows = 0
          @num_output_rows = 0
          @num_rejected_rows = 0
          @outer_thread = Thread.current
          @thread_active = false
          @progress_log_timer = Time.now
          @previous_num_input_rows = 0

          case task['compress']
          when 'GZIP'
            @write_proc = self.method(:write_gzip)
          else
            @write_proc = self.method(:write_uncompressed)
          end
        end

        def thread_dump
          unless $embulk_output_vertica_thread_dumped
            $embulk_output_vertica_thread_dumped = true
            Embulk.logger.debug "embulk-output-vertica: kill -3 #{$$} (Thread dump)"
            begin
              Process.kill :QUIT, $$
            rescue SignalException
            ensure
              sleep 1
            end
          end
        end

        def enqueue(json_page)
          if @thread_active and @thread.alive?
            Embulk.logger.trace { "embulk-output-vertica: enqueue" }
            @queue.push(json_page)
          else
            Embulk.logger.info { "embulk-output-vertica: thread is dead, but still trying to enqueue" }
            thread_dump
            raise RuntimeError, "embulk-output-vertica: thread is died, but still trying to enqueue"
          end
        end

        def write_gzip(io, page, &block)
          buf = Zlib::Deflate.new
          write_buf(buf, page, &block)
          write_io(io, buf.finish)
        end

        def write_uncompressed(io, page, &block)
          buf = ''
          write_buf(buf, page, &block)
          write_io(io, buf)
        end

        PIPE_BUF = 4096

        def write_io(io, str)
          str = str.force_encoding('ASCII-8BIT')
          i = 0
          # split str not to be blocked (max size of pipe buf is 64k bytes on Linux, Mac at default)
          while substr = str[i, PIPE_BUF]
            Embulk.logger.trace { "embulk-output-vertica: io.write with write_timeout:#{@task['write_timeout']}" }
            Timeout.timeout(@task['write_timeout'], WriteTimeoutError) { io.write(substr) }
            i += PIPE_BUF
          end
        end

        def write_buf(buf, json_page, &block)
          json_page.each do |record|
            yield(record) if block_given?
            Embulk.logger.trace { "embulk-output-vertica: record #{record}" }
            buf << record << "\n"
            @num_input_rows += 1
          end
          now = Time.now
          if @progress_log_timer < now - 10 # once in 10 seconds
            speed = ((@num_input_rows - @previous_num_input_rows) / (now - @progress_log_timer).to_f).round(1)
            @progress_log_timer = now
            @previous_num_input_rows = @num_input_rows
            Embulk.logger.info { "embulk-output-vertica: num_input_rows #{num_format(@num_input_rows)} (#{num_format(speed)} rows/sec)" }
          end
        end

        def num_format(number)
          number.to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\1,')
        end

        # @return [Array] dequeued json_page
        # @return [String] 'finish' is dequeued to finish
        def dequeue
          json_page = nil
          Embulk.logger.trace { "embulk-output-vertica: @queue.pop with dequeue_timeout:#{@task['dequeue_timeout']}" }
          Timeout.timeout(@task['dequeue_timeout'], DequeueTimeoutError) { json_page = @queue.pop }
          Embulk.logger.trace { "embulk-output-vertica: dequeued" }
          Embulk.logger.debug { "embulk-output-vertica: dequeued finish" } if json_page == 'finish'
          json_page
        end

        def copy(jv, sql, &block)
          Embulk.logger.debug "embulk-output-vertica: copy, waiting a first message"

          num_output_rows = 0; rejected_row_nums = []; last_record = nil

          json_page = dequeue
          return [num_output_rows, rejected_row_nums, last_record] if json_page == 'finish'

          Embulk.logger.debug "embulk-output-vertica: #{sql}"

          num_output_rows, rejected_row_nums = jv.copy(sql) do |stdin, stream|
            @write_proc.call(stdin, json_page) {|record| last_record = record }

            while true
              json_page = dequeue
              break if json_page == 'finish'
              @write_proc.call(stdin, json_page) {|record| last_record = record }
            end
          end

          @num_output_rows += num_output_rows
          @num_rejected_rows += rejected_row_nums.size
          Embulk.logger.info { "embulk-output-vertica: COMMIT!" }
          jv.commit
          Embulk.logger.debug { "embulk-output-vertica: COMMITTED!" }

          if rejected_row_nums.size > 0
            Embulk.logger.debug { "embulk-output-vertica: rejected_row_nums: #{rejected_row_nums}" }
          end

          [num_output_rows, rejected_row_nums, last_record]
        end

        def run
          Embulk.logger.debug { "embulk-output-vertica: thread started" }
          begin
            jv = Vertica.connect(@task)
            begin
              num_output_rows, rejected_row_nums, last_record = copy(jv, copy_sql)
              Embulk.logger.debug { "embulk-output-vertica: thread finished" }
            rescue java.sql.SQLDataException => e
              if @task['reject_on_materialized_type_error'] and e.message =~ /Rejected by user-defined parser/
                Embulk.logger.warn "embulk-output-vertica: ROLLBACK! some of column types and values types do not fit #{rejected_row_nums}"
              else
                Embulk.logger.warn "embulk-output-vertica: ROLLBACK! #{rejected_row_nums}"
              end
              Embulk.logger.info { "embulk-output-vertica: last_record: #{last_record}" }
              rollback(jv)
              raise e
            rescue => e
              Embulk.logger.warn "embulk-output-vertica: ROLLBACK! #{e.class} #{e.message} #{e.backtrace.join("\n  ")}"
              rollback(jv)
              raise e
            end
          ensure
            close(jv)
          end
        rescue TimeoutError => e
          Embulk.logger.error "embulk-output-vertica: UNKNOWN TIMEOUT!! #{e.class}"
          @thread_active = false # not to be enqueued any more
          dequeue_all
          thread_dump
          exit(1)
        rescue Exception => e
          Embulk.logger.error "embulk-output-vertica: UNKNOWN ERROR! #{e.class} #{e.message} #{e.backtrace.join("\n  ")}"
          @thread_active = false # not to be enqueued any more
          dequeue_all
          @outer_thread.raise e
        end

        def dequeue_all
          Embulk.logger.debug "embulk-output-vertica: dequeue all"
          while @queue.size > 0
            @queue.pop # dequeue all because some might be still trying @queue.push and get blocked, need to release
          end
        end

        def close(jv)
          begin
            jv.close
          rescue java.sql.SQLException => e # The connection is closed
            Embulk.logger.debug "embulk-output-vertica: #{e.class} #{e.message}"
          end
        end

        def rollback(jv)
          begin
            jv.rollback
          rescue java.sql.SQLException => e # The connection is closed
            Embulk.logger.debug "embulk-output-vertica: #{e.class} #{e.message}"
          end
        end

        def start
          @thread = Thread.new(&method(:run))
          @thread_active = true
        end

        def commit
          Embulk.logger.debug "embulk-output-vertica: output_thread commit"
          @thread_active = false
          success = true
          if @thread.alive?
            Embulk.logger.debug { "embulk-output-vertica: push finish with finish_timeout:#{@task['finish_timeout']}" }
            @queue.push('finish')
            Thread.pass
            @thread.join(@task['finish_timeout'])
            if @thread.alive?
              @thread.kill
              Embulk.logger.error "embulk-output-vertica: finish_timeout #{@task['finish_timeout']}sec exceeded, thread is killed forcely"
              success = false
            end
          else
            Embulk.logger.error "embulk-output-vertica: thread died accidently"
            success = false
          end

          task_report = {
            'num_input_rows' => @num_input_rows,
            'num_output_rows' => @num_output_rows,
            'num_rejected_rows' => @num_rejected_rows,
            'success' => success
          }
        end

        # private

        def copy_sql
          @copy_sql ||= "COPY #{quoted_schema}.#{quoted_temp_table} FROM STDIN#{compress}#{fjsonparser}#{copy_mode}#{abort_on_error} NO COMMIT"
        end

        def quoted_schema
          ::Jvertica.quote_identifier(@task['schema'])
        end

        def quoted_table
          ::Jvertica.quote_identifier(@task['table'])
        end

        def quoted_temp_table
          ::Jvertica.quote_identifier(@task['temp_table'])
        end

        def compress
          " #{@task['compress']}"
        end

        def copy_mode
          " #{@task['copy_mode']}"
        end

        def abort_on_error
          @task['abort_on_error'] ? ' ABORT ON ERROR' : ''
        end

        def fjsonparser
          " PARSER fjsonparser(#{reject_on_materialized_type_error})"
        end

        def reject_on_materialized_type_error
          @task['reject_on_materialized_type_error'] ? 'reject_on_materialized_type_error=true' : ''
        end
      end
    end
  end
end


lib/embulk/output/vertica.rb
require 'jvertica'
require 'securerandom'
require_relative 'vertica/value_converter_factory'
require_relative 'vertica/output_thread'

module Embulk
  module Output
    class Vertica < OutputPlugin
      Plugin.register_output('vertica', self)

      class Error < StandardError; end
      class NotSupportedType < Error; end

      def self.thread_pool
        @thread_pool ||= @thread_pool_proc.call
      end

      def self.transaction_report(jv, task, task_reports)
        quoted_schema     = ::Jvertica.quote_identifier(task['schema'])
        quoted_temp_table = ::Jvertica.quote_identifier(task['temp_table'])

        num_input_rows = task_reports.map {|report| report['num_input_rows'].to_i }.inject(:+)
        num_response_rows = task_reports.map {|report| report['num_output_rows'].to_i }.inject(:+)
        result = query(jv, %[SELECT COUNT(*) FROM #{quoted_schema}.#{quoted_temp_table}])
        num_output_rows = result.map {|row| row.values }.flatten.first.to_i
        num_rejected_rows = num_input_rows - num_output_rows
        transaction_report = {
          'num_input_rows' => num_input_rows,
          'num_response_rows' => num_response_rows,
          'num_output_rows' => num_output_rows,
          'num_rejected_rows' => num_rejected_rows,
        }
      end

      def self.transaction(config, schema, task_count, &control)
        task = {
          'host'             => config.param('host',             :string,  :default => 'localhost'),
          'port'             => config.param('port',             :integer, :default => 5433),
          'user'             => config.param('user',             :string,  :default => nil),
          'username'         => config.param('username',         :string,  :default => nil), # alias to :user for backward compatibility
          'password'         => config.param('password',         :string,  :default => ''),
          'database'         => config.param('database',         :string,  :default => 'vdb'),
          'schema'           => config.param('schema',           :string,  :default => 'public'),
          'table'            => config.param('table',            :string),
          'mode'             => config.param('mode',             :string,  :default => 'insert'),
          'copy_mode'        => config.param('copy_mode',        :string,  :default => 'AUTO'),
          'abort_on_error'   => config.param('abort_on_error',   :bool,    :default => false),
          'compress'         => config.param('compress',         :string,  :default => 'UNCOMPRESSED'),
          'default_timezone' => config.param('default_timezone', :string, :default => 'UTC'),
          'column_options'   => config.param('column_options',   :hash,    :default => {}),
          'json_payload'     => config.param('json_payload',     :bool,    :default => false),
          'resource_pool'    => config.param('resource_pool',    :string,  :default => nil),
          'reject_on_materialized_type_error' => config.param('reject_on_materialized_type_error', :bool, :default => false),
          'pool'             => config.param('pool',             :integer, :default => task_count),
          'write_timeout'    => config.param('write_timeout',    :integer, :default => nil), # like 11 * 60 sec
          'dequeue_timeout'  => config.param('dequeue_timeout',  :integer, :default => nil), # like 13 * 60 sec
          'finish_timeout'   => config.param('finish_timeout',   :integer, :default => nil), # like 3 * 60 sec
        }

        @thread_pool_proc = Proc.new do
          OutputThreadPool.new(task, schema, task['pool'])
        end

        task['user'] ||= task['username']
        unless task['user']
          raise ConfigError.new 'required field "user" is not set'
        end

        task['mode'] = task['mode'].upcase
        unless %w[INSERT REPLACE DROP_INSERT].include?(task['mode'])
          raise ConfigError.new "`mode` must be one of INSERT, REPLACE, DROP_INSERT"
        end

        task['copy_mode'] = task['copy_mode'].upcase
        unless %w[AUTO DIRECT TRICKLE].include?(task['copy_mode'])
          raise ConfigError.new "`copy_mode` must be one of AUTO, DIRECT, TRICKLE"
        end

        # ToDo: Support BZIP, LZO
        task['compress'] = task['compress'].upcase
        unless %w[GZIP UNCOMPRESSED].include?(task['compress'])
          raise ConfigError.new "`compress` must be one of GZIP, UNCOMPRESSED"
        end

        now = Time.now
        unique_name = SecureRandom.uuid
        task['temp_table'] = "#{task['table']}_LOAD_TEMP_#{unique_name}"

        quoted_schema     = ::Jvertica.quote_identifier(task['schema'])
        quoted_table      = ::Jvertica.quote_identifier(task['table'])
        quoted_temp_table = ::Jvertica.quote_identifier(task['temp_table'])

        connect(task) do |jv|
          unless task['json_payload'] # ToDo: auto table creation is not supported to json_payload mode yet
            sql_schema_table = self.sql_schema_from_embulk_schema(schema, task['column_options'])

            # create the target table
            query(jv, %[DROP TABLE IF EXISTS #{quoted_schema}.#{quoted_table}]) if task['mode'] == 'DROP_INSERT'
            query(jv, %[CREATE TABLE IF NOT EXISTS #{quoted_schema}.#{quoted_table} (#{sql_schema_table})])
          end

          # create a temp table
          query(jv, %[DROP TABLE IF EXISTS #{quoted_schema}.#{quoted_temp_table}])

          if task['mode'] == 'REPLACE'
            # In the case of replace mode, this temp table is replaced with the original table. So, projections should also be copied
            query(jv, %[CREATE TABLE #{quoted_schema}.#{quoted_temp_table} LIKE #{quoted_schema}.#{quoted_table} INCLUDING PROJECTIONS])
          else
            query(jv, %[CREATE TABLE #{quoted_schema}.#{quoted_temp_table} LIKE #{quoted_schema}.#{quoted_table}])
            # Create internal vertica projection beforehand, otherwirse, parallel copies lock table to create a projection and we get S Lock error sometimes
            # This is a trick to create internal vertica projection
            query(jv, %[INSERT INTO #{quoted_schema}.#{quoted_temp_table} SELECT * FROM #{quoted_schema}.#{quoted_table} LIMIT 0])
          end
          Embulk.logger.trace {
            result = query(jv, %[SELECT EXPORT_OBJECTS('', '#{task['schema']}.#{task['temp_table']}')])
            # You can see `CREATE PROJECTION` if the table has a projection
            "embulk-output-vertica: #{result.to_a.flatten}"
          }
        end

        begin
          # insert data into the temp table
          thread_pool.start
          yield(task)
          task_reports = thread_pool.commit
          Embulk.logger.info { "embulk-output-vertica: task_reports: #{task_reports.to_json}" }

          connect(task) do |jv|
            transaction_report = self.transaction_report(jv, task, task_reports)
            Embulk.logger.info { "embulk-output-vertica: transaction_report: #{transaction_report.to_json}" }

            if task['abort_on_error'] # double-meaning, also used for COPY statement
              if transaction_report['num_input_rows'] != transaction_report['num_output_rows']
                raise Error, "ABORT: `num_input_rows (#{transaction_report['num_input_rows']})` and " \
                  "`num_output_rows (#{transaction_report['num_output_rows']})` does not match"
              end
            end

            if task['mode'] == 'REPLACE'
              # swap table and drop the old table
              quoted_old_table = ::Jvertica.quote_identifier("#{task['table']}_LOAD_OLD_#{unique_name}")
              from = "#{quoted_schema}.#{quoted_table},#{quoted_schema}.#{quoted_temp_table}"
              to   = "#{quoted_old_table},#{quoted_table}"
              query(jv, %[ALTER TABLE #{from} RENAME TO #{to}])
              query(jv, %[DROP TABLE #{quoted_schema}.#{quoted_old_table}])
            else
              # insert select from the temp table
              hint = '/*+ direct */ ' if task['copy_mode'] == 'DIRECT' # I did not prepare a specific option, does anyone want?
              query(jv, %[INSERT #{hint}INTO #{quoted_schema}.#{quoted_table} SELECT * FROM #{quoted_schema}.#{quoted_temp_table}])
              jv.commit
            end
          end
        ensure
          connect(task) do |jv|
            # clean up the temp table
            query(jv, %[DROP TABLE IF EXISTS #{quoted_schema}.#{quoted_temp_table}])
            Embulk.logger.trace { "embulk-output-vertica: select result\n#{query(jv, %[SELECT * FROM #{quoted_schema}.#{quoted_table} LIMIT 10]).map {|row| row.to_h }.join("\n") rescue nil}" }
          end
        end
        # this is for -o next_config option, add some paramters for next time execution if wants
        next_config_diff = {}
        return next_config_diff
      end

      # instance is created on each thread
      def initialize(task, schema, index)
        super
      end

      # called for each page in each thread
      def close
      end

      # called for each page in each thread
      def add(page)
        self.class.thread_pool.enqueue(page)
      end

      def finish
      end

      def abort
      end

      # called after processing all pages in each thread
      # we do commit on #transaction for all pools, not at here
      def commit
        {}
      end

      private

      def self.connect(task)
        jv = ::Jvertica.connect({
          host: task['host'],
          port: task['port'],
          user: task['user'],
          password: task['password'],
          database: task['database'],
        })

        if resource_pool = task['resource_pool']
          query(jv, "SET SESSION RESOURCE_POOL = #{::Jvertica.quote(resource_pool)}")
        end

        if block_given?
          begin
            yield jv
          ensure
            jv.close
          end
        end
        jv
      end

      # @param [Schema] schema embulk defined column types
      # @param [Hash]   column_options user defined column types
      # @return [String] sql schema used to CREATE TABLE
      def self.sql_schema_from_embulk_schema(schema, column_options)
        sql_schema = schema.names.zip(schema.types).map do |column_name, type|
          if column_options[column_name] and column_options[column_name]['type']
            sql_type = column_options[column_name]['type']
          else
            sql_type = sql_type_from_embulk_type(type)
          end
          [column_name, sql_type]
        end
        sql_schema.map {|name, type| "#{::Jvertica.quote_identifier(name)} #{type}" }.join(',')
      end

      def self.sql_type_from_embulk_type(type)
        case type
        when :boolean then 'BOOLEAN'
        when :long then 'INT' # BIGINT is a synonym for INT in vertica
        when :double then 'FLOAT' # DOUBLE PRECISION is a synonym for FLOAT in vertica
        when :string then 'VARCHAR' # LONG VARCHAR is not recommended. Default is VARCHAR(80)
        when :timestamp then 'TIMESTAMP'
        else raise NotSupportedType, "embulk-output-vertica cannot take column type #{type}"
        end
      end

      def self.query(conn, sql)
        Embulk.logger.info "embulk-output-vertica: #{sql}"
        conn.query(sql)
      end

      def query(conn, sql)
        self.class.query(conn, sql)
      end
    end
  end
end
